tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c

```diff
@@ -0,0 +1,457 @@
+	__asm__ volatile (
+		"mov r1, r7;\n\t"
+		"ldmia %0, {r4-r11};\n\t"
+		"mov r7, r1;\n\t"
+		:
+		: "r" (regs)
+		: "memory"
+	);
+	__DSB();
+}
+
+static void verify_callee_saved(const _callee_saved_t *src,
+		const _callee_saved_t *dst)
+{
+	/* Verify callee-saved registers are as expected */
+	zassert_true((src->v1 == dst->v1)
+			&& (src->v2 == dst->v2)
+			&& (src->v3 == dst->v3)
+			&& (src->v4 == dst->v4)
+			&& (src->v5 == dst->v5)
+			&& (src->v6 == dst->v6)
+			&& (src->v7 == dst->v7)
+			&& (src->v8 == dst->v8),
+		" got: 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x\n"
+		" expected:  0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x\n",
+		src->v1,
+		src->v2,
+		src->v3,
+		src->v4,
+		src->v5,
+		src->v6,
+		src->v7,
+		src->v8,
+		dst->v1,
+		dst->v2,
+		dst->v3,
+		dst->v4,
+		dst->v5,
+		dst->v6,
+		dst->v7,
+		dst->v8
+	);
+}
+
+#if defined(CONFIG_FLOAT) && defined(CONFIG_FP_SHARING)
+
+struct _preempt_float ztest_thread_fp_callee_saved_regs = {
+	.s16 = 0x16, .s17 = 0x17, .s18 = 0x18, .s19 = 0x19,
+	.s20 = 0x20, .s21 = 0x21, .s22 = 0x22, .s23 = 0x23,
+	.s24 = 0x24, .s25 = 0x25, .s26 = 0x26, .s27 = 0x27,
+	.s28 = 0x28, .s29 = 0x29, .s30 = 0x30, .s31 = 0x31,
+};
+
+static void load_fp_callee_saved_regs(
+	const volatile struct _preempt_float *regs)
+{
+	__asm__ volatile (
+		"vldmia %0, {s16-s31};\n\t"
+		:
+		: "r" (regs)
+		: "memory"
+		);
+	__DSB();
+}
+
+static void verify_fp_callee_saved(const struct _preempt_float *src,
+		const struct _preempt_float *dst)
+{
+	/* Verify FP callee-saved registers are as expected */
+	zassert_true((src->s16 == dst->s16)
+			&& (src->s17 == dst->s17)
+			&& (src->s18 == dst->s18)
+			&& (src->s19 == dst->s19)
+			&& (src->s20 == dst->s20)
+			&& (src->s21 == dst->s21)
+			&& (src->s22 == dst->s22)
+			&& (src->s23 == dst->s23)
+			&& (src->s24 == dst->s24)
+			&& (src->s25 == dst->s25)
+			&& (src->s26 == dst->s26)
+			&& (src->s27 == dst->s27)
+			&& (src->s28 == dst->s28)
+			&& (src->s29 == dst->s29)
+			&& (src->s30 == dst->s30)
+			&& (src->s31 == dst->s31),
+		" got: 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x"
+		" 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x\n"
+		" expected:  0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x"
+		" 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x 0x%0x\n",
+		src->s16,
+		src->s17,
+		src->s18,
+		src->s19,
+		src->s20,
+		src->s21,
+		src->s22,
+		src->s23,
+		src->s24,
+		src->s25,
+		src->s26,
+		src->s27,
+		src->s28,
+		src->s29,
+		src->s30,
+		src->s31,
+		dst->s16,
+		dst->s17,
+		dst->s18,
+		dst->s19,
+		dst->s20,
+		dst->s21,
+		dst->s22,
+		dst->s23,
+		dst->s24,
+		dst->s25,
+		dst->s26,
+		dst->s27,
+		dst->s28,
+		dst->s29,
+		dst->s30,
+		dst->s31
+	);
+}
+
+#endif /* CONFIG_FLOAT && CONFIG_FP_SHARING */
+
+static void alt_thread_entry(void)
+{
+	zassert_true(switch_flag == 0,
+		"Alternative thread: switch flag not zero on thread entry\n");
+
+	/* Increment switch flag */
+	switch_flag = 1;
+
+#if defined(CONFIG_NO_OPTIMIZATIONS)
```

Comment:
Why this check is under CONFIG_NO_OPTIMIZATIONS? Do we have there dependency to the code located somewhere else? If so, could we have short comment here?

Reply:
Because of the way the SWAP is triggered on the main test thread
- under !NO_OPTIMIZATIONS, I call swap(_basepri_)
- under NO_OPTIMIZATIONS, I set the PendSV interrupt directly

We can test both stuff, in this way :) Just that we need to be careful what we ASSERT on, in the alternative thread.